src/core/searcher.rs

```diff
@@ -112,6 +132,72 @@ impl Searcher {
         store_reader.get_async(doc_address.doc_id).await
     }
 
+    /// Fetches a set of documents in an asynchronous manner.
+    ///
+    /// This method is more efficient than calling [`doc_async`](Self::doc_async) multiple times, as
+    /// it batches overlapping requests to segments and blocks.
+    #[cfg(feature = "quickwit")]
+    pub async fn docs_async<D: DocumentDeserialize>(
```

Comment:
I understand that this is not a particularly technical objection, but I do wonder whether it is helpful to add this as additional API surface to Tantivy itself. User code can already sort `DocAddress` and dispatch multiple tasks and/or futures for each segment. And user code usually has the best understanding which trade-off to take w.r.t. parallelism, performance and locality.

For example, a service serving multiple users would probably aim for increased locality by handling a single user request as a single task accessing documents in address order to improve cache efficiency. Such a service would handle parallelism and concurrency by handling multiple users in multiple tasks.

Furthermore, I have a hard time see a situation where additional within-segment concurrency  is better than serial cache-friendly access patterns. Also, the additional data structure required to track this additional concurrency seem ripe with trade-offs which user code is just in a better position to make, e.g. use `FuturesUnordered` to concurrency but not parallelism or spawn multiple tasks into a `JoinSet` to actually load multiple hardware threads.

Long story short, personally, I am not convinced this API is pulling its weight in terms of complexity and implied trade-offs.

Reply:
I would say that main problem is that user code lacks information required to choose the best access pattern. For example, a function which informs user code about the blocks for a given document (opaquely like `fn cache_key(DocId) -> usize`) and a way to clone a `StoreReader` (to avoid reopening from disk, but get independent block caches) might suffice to know which documents end up in the same block and to control which I/O is issued against the same cache instance.

(I guess a function which takes something like `&mut [(DocId, usize)]` and sorts that slice and with the cache keys filled in might be nice to get the most computational efficiency, but I suspect the cost of repeatedly calling `block_checkpoint` is still dwarfed by the cost of I/O and decompression. Alternatively `&mut BTreeMap<DocId, usize>` would be more inline with what is proposed here.)

Reply:
> Maybe it's a function to fetch multiple docs at once (originally my suggestion) but that makes tantivy choose a tradeoff which might not be the best for your applicatin.

While I agree that it may be beneficial to have the fine-grained control to choose the trade-off, from an API design perspective, I believe default behavior should be that users don't have to know about these implementation details. This should be an opt-in API. Otherwise, it'd be nice if the non-fine-tuned path had sane, reasonably performant defaults.

> If you sort your docs, and fetch multiple docs concurrently, you might end up fetching multiple docs from the same block, concurrently.
This is very good to know. At least in this change, I try to group all documents within the same block into one future to minimize this. Of course, multiple separate calls will not benefit from this grouping but I'd want to see some cases where that is a common-enough pattern to warrant further optimization.

Reply:
> This is very good to know. At least in this change, I try to group all documents within the same block into one future to minimize this. Of course, multiple separate calls will not benefit from this grouping but I'd want to see some cases where that is a common-enough pattern to warrant further optimization.

I think the point is here exactly that those futures run concurrently against the same cache and hence will evict each others blocks. After all, fetching multiple blocks concurrently is sort of the point of concurrency here. Hence, I think we should start with cheaply cloning `StoreReader` to give the programs control over caches (and the function to opaquely inform user code about cache keys) and then maybe use this here if we can reach a compromise w.r.t. concurrency/parallelism/etc. for a default API.

(Concurrency in contrast to parallelism likely does little in a normal Tantivy local-disk-based setup in contrast to a directory implementation calling into e.g. S3. Parallelism on the other hand would also benefit CPU-bound decompression but usually needs integration with the runtime to spawn tasks.)

Reply:
I proposed some preliminary API additions in #2276. Using these, I think fetching multiple documents should consistent of
* sort all documents into groups with the same cache key/from the same block
* deciding how many independent block caches should be used and how large those should be
* then futures should pull groups of documents and handle them using their private cache instance
* these futures should either be spawn onto independent tasks (parallelism, e.g. `JoinSet`) or polled concurrently (e.g. `FuturesUnordered`)
* ideally it would be possible to reuse those store readers (and spawned tasks) to fetch multiple groups of documents

Obviously there are quite a few trade-offs implied here, the big one being concurrency versus parallelism. Some can possibly be side stepped, e.g always use a size-1 block each for each group and spawn as many futures/tasks as groups which would always avoid conflicts at the cost of more in-memory data structures. Similarly, we could forgo reuse at the same cost of setting up in-memory data structures.

Reply:
Maybe we can side-step the concurrency/parallelism trade-off here by providing an API that just yields the futures and requires the caller to poll them to completion: So a signature like

```rust
async fn docs_async(&self, doc_addresses: impl IntoIterator<Item=DocAddress>) -> impl Iterator<Item=impl Future<Output=crate::Result<Vec<D>>>>;
```

which internally sets up the per-segment-per-block groups (e.g. sorting them into `HashMap<SegmentOrdinal, HashMap<CacheKey, Vec<DocId>>`) with independent size-1 block caches (using e.g. `self.store_readers[segment_ord as usize].with_independent_cache(1)`) and yields a future for each group which the caller can then push into a `FuturesUnordered` or spawn onto a `JoinSet` as they see fit.

Reply:
Sorry for somewhat hijacking this cache, but I pushed an implementation of the above to #2276 which as discussed is still a specific trade-off but should avoid cache trashing while still reusing the existing block cache via `StoreReader::get_async` and leaving the decision of whether to use concurrency or parallelism to the user code (which is very little syntactical overhead if it just ends up being `FuturesUnordered` as the modified test case shows).